File: src/main/kotlin/lunchbox/domain/resolvers/LunchResolverFeldkueche.kt
package lunchbox.domain.resolvers

import java.net.URL
import java.time.DayOfWeek
import java.time.LocalDate

import lunchbox.domain.models.LunchOffer
import lunchbox.domain.models.LunchProvider.FELDKUECHE
import lunchbox.util.html.HtmlParser
import lunchbox.util.ocr.OcrClient
import lunchbox.util.string.StringParser
import org.joda.money.Money
import org.springframework.stereotype.Component

/**
 * Mittagsangebote von Feldküche Rühlow ermitteln.
 */
@Component
class LunchResolverFeldkueche(
  val dateValidator: DateValidator,
  val ocrClient: OcrClient
) : LunchResolver {

  override val provider = FELDKUECHE

  override fun resolve(): List<LunchOffer> =
    // TODO: parallelize
    resolveFromImageLinks(resolveImageLinks(provider.menuUrl))

  fun resolveImageLinks(htmlUrl: URL): List<URL> {
    val site = HtmlParser.parse(htmlUrl)

    val divsWithContent = site.select("div#content_area")
    val links = divsWithContent.select("a").map { it.attr("href") }
    return links.map { URL(it) }
  }

  private fun resolveFromImageLinks(imageUrls: List<URL>): List<LunchOffer> =
    imageUrls.flatMap { resolveTextFromImageLink(it) }

  private fun resolveTextFromImageLink(imageUrl: URL): List<LunchOffer> =
    resolveOffersFromText(ocrClient.doOCR(imageUrl))

  fun resolveOffersFromText(rawText: String): List<LunchOffer> {
    val contentAsLines = rawText.split('\n').map { correctOcrErrors(it) }
    val monday = resolveMonday(contentAsLines) ?: return emptyList()

    val rawOffers =
      if (Weekday.values().any { contentAsLines.contains(it.label) })
        resolveOffersWith3rowSplit(contentAsLines)
      else
        resolveOffersWith2rowSplit(contentAsLines)

    return rawOffers.map {
      LunchOffer(0, it.name, monday.plusDays(it.weekday.order), it.price, provider.id)
    }
  }

  private fun resolveMonday(contentAsLines: List<String>): LocalDate? {
    val wochenplanZeile = contentAsLines.find { it.contains("Wochenplan") } ?: return null
    val day = StringParser.parseLocalDate(wochenplanZeile) ?: return null
    val monday = day.with(DayOfWeek.MONDAY)
    return if (dateValidator.isValid(monday)) monday else null
  }

  private fun resolveOffersWith2rowSplit(contentAsLines: List<String>): List<RawOffer> {
    val weekdaysRegex = Regex("(${Weekday.values().joinToString("|"){ it.label }}) .*")
    val offerTexts =
      contentAsLines
        .filter { it.matches(weekdaysRegex) }
        .filterNot { it.startsWith("Montag bis") }

    fun splitWeekdayAndName(text: String): RawOfferName? {
      val weekdayName = text.takeWhile { it != ' ' }
      val lunchName = text.dropWhile { it != ' ' }
      val weekday = Weekday.values().find { it.label == weekdayName } ?: return null

      return RawOfferName(weekday, lunchName.replace(Regex(""" \d+,\d{2} *€"""), "").trim())
    }

    val prices = contentAsLines
      .filter { it.matches(Regex(""".*\d+,(\d{2}) *€""")) }
      .mapNotNull { StringParser.parseMoney(it) }

    return offerTexts
      .mapNotNull { splitWeekdayAndName(it) }
      .zip(prices)
      .map { (rawName, price) -> RawOffer(rawName.weekday, rawName.name, price) }
  }

  private fun resolveOffersWith3rowSplit(contentAsRows: List<String>): List<RawOffer> {
    fun isWeekday(text: String) = Weekday.values().map { it.label }.contains(text)
    val relevantRows =
      contentAsRows
        .filter { it.isNotEmpty() }
        .dropWhile { !isWeekday(it) }
    val weekdayTextRows = relevantRows.takeWhile { isWeekday(it) }
    val followingRows = relevantRows.dropWhile { isWeekday(it) }

    val weekdayRows =
      weekdayTextRows
        .mapNotNull { row -> Weekday.values().find { it.label == row } }

    val prices =
      followingRows
        .filter { it.matches(Regex(""".*\d+,(\d{2}) *€""")) }
        .mapNotNull { StringParser.parseMoney(it) }

    return weekdayRows
      .zip(followingRows)
      .map { (weekday, name) -> RawOfferName(weekday, name) }
      .zip(prices)
      .map { (rawName, price) -> RawOffer(rawName.weekday, rawName.name, price) }
  }

  private fun correctOcrErrors(line: String) =
    line.trim()
      .replace("Wochenglan", "Wochenplan")
      .replace("‚", ",")
      .replace("""l)""", "0")
      .replace(Regex("""(\d+,\d{2}) [cCeEs]"""), "$1 €")
      //      .replace("4,00 €", "4,80 €") // besser falsch zu hohe 4,80 als zu niedrige 4,00
      .replace("Kanonen-", "Kartoffeln")
      .replace("ﬂ", "fl")
      .replace("ﬁ", "fi")
      .replace(Regex("Wochen.lan"), "Wochenplan")
      .replace(Regex("Wochen.Ian"), "Wochenplan")
      .replace("eintopl", "eintopf")
      .replace("eintop", "eintopf")
      .replace("eintof", "eintopf")
      .replace("eintopff", "eintopf")
      .replace("Erhsen", "Erbsen")
      .replace("Chili con Garne", "Chili con Carne")
      .replace("—", "-")

  data class RawOfferName(
    val weekday: Weekday,
    val name: String
  )

  data class RawOffer(
    val weekday: Weekday,
    val name: String,
    val price: Money
  )

  enum class Weekday(
    val label: String,
    val order: Long
  ) {
    MONTAG("Montag", 0),
    DIENSTAG("Dienstag", 1),
    MITTWOCH("Mittwoch", 2),
    DONNERSTAG("Donnerstag", 3),
    FREITAG("Freitag", 4);
  }
}

File: backend-play-akka-scala/app/domain/logic/LunchResolverFeldkueche.scala
package domain.logic

import java.net.URL
import java.time.{DayOfWeek, LocalDate}
import java.time.format.DateTimeFormatter

import domain.models.{LunchOffer, LunchProvider}
import domain.util.Html
import domain.util.Html._
import infrastructure.OcrClient
import org.joda.money.{CurrencyUnit, Money}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.matching.Regex

/**
 * Mittagsangebote von Feldküche Rühlow ermitteln.
 */
class LunchResolverFeldkueche(
    dateValidator: DateValidator,
    ocrClient: OcrClient) extends LunchResolver {

  sealed abstract class Weekday(val name: String, val order: Int)

  object Weekday {
    case object MONTAG extends Weekday("Montag", 0)
    case object DIENSTAG extends Weekday("Dienstag", 1)
    case object MITTWOCH extends Weekday("Mittwoch", 2)
    case object DONNERSTAG extends Weekday("Donnerstag", 3)
    case object FREITAG extends Weekday("Freitag", 4)

    val weekdaysValues = List[Weekday](MONTAG, DIENSTAG, MITTWOCH, DONNERSTAG, FREITAG)
    // TODO: improve with macro, see https://github.com/d6y/enumeration-examples & http://underscore.io/blog/posts/2014/09/03/enumerations.html
    val values = weekdaysValues
  }

  override def resolve: Future[Seq[LunchOffer]] =
    Future { resolveImageLinks(new URL("https://www.feldkuechebkarow.de/speiseplan")) }
      .flatMap(imageUrls => resolveFromImageLinks(imageUrls))

  private[logic] def resolveImageLinks(htmlUrl: URL): Seq[URL] = {
    val siteAsXml = Html.load(htmlUrl)

    val divsWithContent = (siteAsXml \\ "div") filter hasId("content_area")
    val links = (divsWithContent \\ "a").map(_ \@ "href")
    links.map { link => new URL(link) }.toSeq
  }

  private def resolveFromImageLinks(imageUrls: Seq[URL]): Future[Seq[LunchOffer]] = Future.sequence {
    imageUrls.map(imageUrl => resolveTextFromImageLink(imageUrl))
  }.map(_.flatten)

  private def resolveTextFromImageLink(imageUrl: URL): Future[Seq[LunchOffer]] =
    ocrClient.doOCR(imageUrl).map(resolveOffersFromText)

  private[logic] def resolveOffersFromText(rawText: String): Seq[LunchOffer] = {
    val contentAsLines = rawText.split('\n').map(correctOcrErrors)
    val mondayOpt = resolveMonday(contentAsLines)

    val rawOffers =
      if (Weekday.values.exists(weekday => contentAsLines.contains(weekday.name)))
        resolveOffersWith3rowSplit(contentAsLines)
      else
        resolveOffersWith2rowSplit(contentAsLines)

    for (
      (weekday, name, price) <- rawOffers;
      monday <- mondayOpt
    ) yield LunchOffer(0, name, monday.plusDays(weekday.order), price, LunchProvider.FELDKUECHE.id)
  }

  private def resolveMonday(contentAsLines: Seq[String]): Option[LocalDate] =
    for (
      wochenplanZeile <- contentAsLines.find(_.contains("Wochenplan"));
      day <- parseDay(wochenplanZeile);
      monday = day.`with`(DayOfWeek.MONDAY) if dateValidator.isValid(monday)
    ) yield monday

  private def resolveOffersWith2rowSplit(contentAsLines: Seq[String]): Seq[(Weekday, String, Money)] = {
    val weekdaysRegex = s"(${Weekday.values.map(_.name).mkString("|")}) .*"
    val offerTexts = contentAsLines.filter(_.matches(weekdaysRegex)).filterNot(_.startsWith("Montag bis"))

    def splitWeekdayAndName(text: String): Option[(Weekday, String)] = {
      val (weekdayName, lunchName) = text.span(_ != ' ')
      Weekday.values.find(_.name == weekdayName) match {
        case Some(weekday) => Some(weekday, lunchName.replaceAll(""" \d{1,},\d{2} €""", "").trim)
        case _ => None
      }
    }

    val prices = contentAsLines.filter(_.matches(""".*\d{1,},(\d{2}) €""")).flatMap(parsePrice)

    offerTexts
      .flatMap(splitWeekdayAndName)
      .zip(prices)
      .map{ case ((weekday, name), price) => (weekday, name, price) }
  }

  private def resolveOffersWith3rowSplit(contentAsRows: Seq[String]): Seq[(Weekday, String, Money)] = {
    def isWeekday(text: String) = Weekday.values.map(_.name).contains(text)
    val (weekdayTextRows, followingRows) =
      contentAsRows
        .filter(_.nonEmpty)
        .dropWhile(line => !isWeekday(line))
        .span(isWeekday)

    val weekdayRows = weekdayTextRows.flatMap(row => Weekday.values.find(_.name == row))
    val prices = followingRows.filter(_.matches(""".*\d{1,},(\d{2}) €""")).flatMap(parsePrice)

    weekdayRows
      .zip(followingRows)
      .zip(prices)
      .map{ case ((weekday, name), price) => (weekday, name, price) }
  }

  private def correctOcrErrors(line: String) =
    line.trim
      .replaceAll("Wochenglan", "Wochenplan")
      .replaceAll("‚", ",")
      .replaceAll("""l\)""", "0")
      .replaceAll("""(\d{1,},\d{2}) [cCeEs]""", "$1 €")
      //      .replaceAll("4,00 €", "4,80 €") // besser falsch zu hohe 4,80 als zu niedrige 4,00
      .replaceAll("Kanonen-", "Kartoffeln")
      .replaceAll("ﬂ", "fl")
      .replaceAll("ﬁ", "fi")
      .replaceAll("Wochen.lan", "Wochenplan")
      .replaceAll("Wochen.Ian", "Wochenplan")
      .replaceAll("eintopl", "eintopf")
      .replaceAll("eintop", "eintopf")
      .replaceAll("eintof", "eintopf")
      .replaceAll("eintopff", "eintopf")
      .replaceAll("Erhsen", "Erbsen")
      .replaceAll("Chili con Garne", "Chili con Carne")
      .replaceAll("—", "-")

  implicit class RegexContext(sc: StringContext) {
    def r = new Regex(sc.parts.mkString, sc.parts.tail.map(_ => "x"): _*)
  }

  /**
   * Erzeugt ein LocalDate aus dem Format "*dd.mm.yyyy*"
   *
   * @param text auszuwertender Text
   * @return
   */
  private def parseDay(text: String): Option[LocalDate] = text match {
    case r""".*(\d{2}\.\d{2}\.\d{2})$dayString.*""" => parseLocalDate(dayString, "dd.MM.yy")
    case _ => None
  }

  /**
   * Erzeugt ein Money-Objekt (in EURO) aus dem Format "*0,00*"
   *
   * @param text auszuwertender Text
   * @return
   */
  private def parsePrice(text: String): Option[Money] = text match {
    case r""".*(\d{1,})$major,(\d{2})$minor €""" => Some(Money.ofMinor(CurrencyUnit.EUR, major.toInt * 100 + minor.toInt))
    case _ => None
  }

  private def parseLocalDate(dateString: String, dateFormat: String): Option[LocalDate] =
    try {
      Some(LocalDate.from(DateTimeFormatter.ofPattern(dateFormat).parse(dateString)))
    } catch {
      case exc: Throwable => None
    }

}
